pkg/workloadmeta/collectors/internal/containerd/image.go

```diff
@@ -0,0 +1,358 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+//go:build containerd
+// +build containerd
+
+package containerd
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/containerd/containerd"
+	"github.com/containerd/containerd/api/events"
+	"github.com/containerd/containerd/content"
+	containerdevents "github.com/containerd/containerd/events"
+	"github.com/containerd/containerd/images"
+	"github.com/containerd/containerd/namespaces"
+	"github.com/gogo/protobuf/proto"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+	"github.com/DataDog/datadog-agent/pkg/workloadmeta"
+)
+
+const imageTopicPrefix = "/images/"
+
+// Events from containerd contain an image name, but not IDs. When a delete
+// event arrives it'll contain a name, but we won't be able to access the ID
+// because the image is already gone, that's why we need to keep the IDs =>
+// names relationships.
+type knownImages struct {
+	// Store IDs and names in both directions for efficient access
+	idsByName map[string]string              // map name => ID
+	namesByID map[string]map[string]struct{} // map ID => set of names
+}
+
+func newKnownImages() *knownImages {
+	return &knownImages{
+		idsByName: make(map[string]string),
+		namesByID: make(map[string]map[string]struct{}),
+	}
+}
+
+func (images *knownImages) addAssociation(imageName string, imageID string) {
+	images.idsByName[imageName] = imageID
+
+	if images.namesByID[imageID] == nil {
+		images.namesByID[imageID] = make(map[string]struct{})
+	}
+	images.namesByID[imageID][imageName] = struct{}{}
+}
```

Comment:
As there’s no lock, the `knownImages` methods are not thread-safe.
Do you confirm that it’s not an issue because all the `knownImages` accesses will be done from a single goroutine, the one that runs the `(c *collector) stream(ctx context.Context)` endless loop ?
I think so, but I’m just asking to be sure.

Reply:
Yes. The containerd events received are handled sequentially in `stream`, so there's no need to make this struct thread-safe.